src/kernels/jupyter/launcher/jupyterExecution.ts

```diff
@@ -182,6 +182,7 @@ export class JupyterExecutionBase implements IJupyterExecution {
                 }
                 throw lastTryError;
             }
+            throw new Error('Max number of attempts reached');
```

Comment:
This function was the root cause, we have an explicit `return undefined` and an implicit at the end of the while loop.
That causes return value to have `undefined` and bubbles all the way the to the top.
This is an unlikely scenario (cancellation is handled higher up, hence we don't need to swallow cancellations and disposing classes here)